plugins/user.py

```diff
@@ -0,0 +1,507 @@
+        have a boto configuration file present that contains the various credentials needed to connect to AWS to perform
+        various actions. The credentials referenced by this profile must be able to be used to upload new keys to the S3
+        bucket specified by TWOFACTOR_S3_BUCKET and generate URLs for those objects.
+
+    The following settings are very sensitive and should be passed in through environment variables:
+
+    TWOFACTOR_SECRET: This is secret key that is used to encrypt sensitive data in redis. If redis is compromised, user
+        tokens would not also be compromised unless this secret was also known.
+    TWOFACTOR_S3_BUCKET: The S3 bucket in which QR code images are stored. This bucket should only be readable by the
+        AWS account which uploads the images. Short lived access will be allowed to the images when they are uploaded to
+        allow the user enough time to retrieve the image.
+    """
+
+    @require_settings("TWOFACTOR_SECRET", "TWOFACTOR_ISSUER", "TWOFACTOR_S3_BUCKET", "TWOFACTOR_S3_PROFILE")
+    @respond_to("^twofactor verify (?P<token>\w+)")
+    @warn_direct_message
+    def verify_user_twofactor(self, message, token):
+        """twofactor verify [token]: start a new authenticated session by providing a verification token generated by the external device"""
+        user = User.get_from_message(self, message)
+        if not user:
+            return
+
+        if user.verify_token(token):
+            self.direct_reply(message, "You are authenticated, your session expires {}".format(self.to_natural_day_and_time(user.session_expiration_time)))
+        else:
+            self.direct_reply(message, "Authentication failed, please try again")
+
+        user.save()
+
+    def direct_reply(self, message, content, **kwargs):
+        self.send_direct_message(message.sender["hipchat_id"], content, **kwargs)
+
+    @respond_to("^twofactor me")
+    @warn_direct_message
+    def create_user_twofactor(self, message):
+        """twofactor me: generate a new QR code that can be used to configure the external device to generate valid verification tokens"""
+        user = User.get(self, message.sender)
+        if not user:
+            user = User.create(self, message.sender)
+            self.direct_reply(message, user.generate_and_upload_qr_code_image(), html=True)
+            user.save()
+            self.direct_reply(message, "Say 'twofactor verify [token]' to start an authenticated session")
+        else:
+            self.direct_reply(message, "twofactor is already configured")
+
+    @respond_to("^twofactor status")
+    @warn_direct_message
+    def twofactor_status(self, message):
+        """twofactor status: tells you if you are currently authenticated and when your session ends"""
+        user = User.get_from_message(self, message)
+        if not user:
+            return
```

Comment:
It does, I pushed the message into User.get_from_message since I needed it pretty much everywhere it was called... not terribly happy about that solution since it "poisons" the model with a weird side-effect. Open to other ideas here...